concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -161,37 +159,46 @@ processFinalization newFinBlock finRec@FinalizationRecord{..} = do
-                       -- by increasing height.
-            pruneTrunk toRemove _ Seq.Empty = return toRemove
-            pruneTrunk toRemove keeper (brs Seq.:|> l) = do
-                parent <- bpParent keeper
-                let toRemove1 = filter (/= keeper) l ++ toRemove
-                toRemove2 <- pruneTrunk toRemove1 parent brs
-                -- mark blocks as finalized now, so that blocks are marked finalized by increasing height
-                markFinalized (getHash keeper) finRec
-                logEvent Skov LLDebug $ "Block " ++ show keeper ++ " marked finalized"
-                -- Finalize the transactions of the surviving block.
-                -- (This is handled in order of finalization.)
-                finalizeTransactions (getHash keeper) (blockSlot keeper) (blockTransactions keeper)
-                ati <- bpTransactionAffectSummaries keeper
-                bcTime <- getSlotTimestamp (blockSlot keeper)
-                let ctx = BlockContext{
-                        bcHash = getHash keeper,
-                        bcHeight = bpHeight keeper,
-                        ..}
-                flushBlockSummaries ctx ati =<< getSpecialOutcomes =<< blockState keeper
-                return toRemove2
-
-        toRemoveFromTrunk <- pruneTrunk [] newFinBlock (Seq.take pruneHeight oldBranches)
+                       -> m ([BlockPointerType m], [BlockPointerType m])
+                       -- ^ The return value is a list of blocks to mark dead (ordered by decreasing
+                       -- height) and a list of blocks to mark finalized (ordered by increasing
+                       -- height).
+            pruneTrunk toRemove toFinalize _ Seq.Empty = return (toRemove, toFinalize)
+            pruneTrunk toRemove toFinalize keeper (brs Seq.:|> l) = do
+              let toRemove1 = toRemove ++ filter (/= keeper) l
+              parent <- bpParent keeper
+              pruneTrunk toRemove1 (keeper : toFinalize) parent brs
+        (toRemoveFromTrunk, toFinalize) <- pruneTrunk [] [] newFinBlock (Seq.take pruneHeight oldBranches)
+        -- Add the finalization to the finalization list
+        addFinalization newFinBlock finRec
+        -- mark blocks as finalized in the order returned by `pruneTrunk`, so that blocks are marked
+        -- finalized by increasing height
+        mfs <- forM toFinalize $ \block -> do
+          markFinalized (getHash block) finRec
+        -- Finalize the transactions of surviving blocks in the order of their finalization.
+        fts <- forM toFinalize $ \block -> do
+          finalizeTransactions (getHash block) (blockSlot block) (blockTransactions block)
+        forM_ toFinalize $ \block -> do
+          ati <- bpTransactionAffectSummaries block
+          bcTime <- getSlotTimestamp (blockSlot block)
+          let ctx = BlockContext{
+                bcHash = getHash block,
+                bcHeight = bpHeight block,
+                ..}
```

Comment:
Is there a reason you are iterating `toFinalize` three times instead of once, doing all three things for each of the blocks?

You could also change the type of wrapupFinalization to take a list of pairs to make this even simpler.

Reply:
I guess the reason was exactly that: I made `wrapupFinalization` take a pair of lists and not a list of pairs, which required the two lists to to be produced independently of each other. And the persistent implementation of `wrapupFinalization` is no more than a wrapper around `writeFinalizationComposite` which I also made to take a pair of lists and not a list of pairs.

So, yeah, there isn't a special reason to do that. I'll change it.